include/mbedtls/mps/writer.h

```diff
@@ -0,0 +1,694 @@
+ * New buffers can be requested through calls to mbedtls_writer_get(),
+ * while previously obtained output buffers can be marked processed
+ * through calls to mbedtls_writer_commit(), emptying the list of
+ * output buffers and invalidating them from the consumer's perspective.
+ *
+ *  Copyright (C) 2006-2015, ARM Limited, All Rights Reserved
+ *  SPDX-License-Identifier: Apache-2.0
+ *
+ *  Licensed under the Apache License, Version 2.0 (the "License"); you may
+ *  not use this file except in compliance with the License.
+ *  You may obtain a copy of the License at
+ *
+ *  http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *  Unless required by applicable law or agreed to in writing, software
+ *  distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ *  WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ *
+ *  This file is part of mbed TLS (https://tls.mbed.org)
+ */
+
+#ifndef MBEDTLS_WRITER_H
+#define MBEDTLS_WRITER_H
+
+#include <stdio.h>
+
+struct mbedtls_writer;
+typedef struct mbedtls_writer mbedtls_writer;
+
+struct mbedtls_writer_ext;
+typedef struct mbedtls_writer_ext mbedtls_writer_ext;
+
+/*
+ * Error codes returned from the writer.
+ */
+
+/** An attempt was made to reclaim a buffer from the writer,
+ *  but the buffer hasn't been fully used up, yet.            */
+#define MBEDTLS_ERR_WRITER_DATA_LEFT             -0x1
+/** The validation of input parameters failed.                */
+#define MBEDTLS_ERR_WRITER_INVALID_ARG           -0x2
+/** The provided outgoing data buffer was not large enough to
+ *  hold all queued data that's currently pending to be
+ *  delivered.                                                */
+#define MBEDTLS_ERR_WRITER_NEED_MORE             -0x3
+/** The requested operation is not possible
+ *  in the current state of the writer.                       */
+#define MBEDTLS_ERR_WRITER_UNEXPECTED_OPERATION  -0x4
+/** The remaining amount of space for outgoing data is not
+ *  sufficient to serve the user's request. The current
+ *  outgoing data buffer must be reclaimed, dispatched,
+ *  and a fresh outgoing data buffer must be fed to the
+ *  writer.                                                   */
+#define MBEDTLS_ERR_WRITER_OUT_OF_DATA           -0x5
+/** A write-request was issued to the extended writer that
+ *  exceeds the bounds of the most recently added group.      */
+#define MBEDTLS_ERR_WRITER_BOUNDS_VIOLATION      -0x9
+/** The extended writer has reached the maximum number of
+ *  groups, and another group cannot be added.                */
+#define MBEDTLS_ERR_WRITER_TOO_MANY_GROUPS       -0xa
+
+/** The identifier to use in mbedtls_writer_reclaim() to
+ *  force the reclamation of the outgoing data buffer even
+ *  if there's space remaining.                               */
+#define MBEDTLS_WRITER_RECLAIM_FORCE 1
```

Comment:
@AndrzejKurek Replying to that old comment: I am slightly reluctant to use `enums` because compilers might realize them through as `int` and waste a few bytes of memory this way.